### PR TITLE
[FN] Exclude transactions with too-new timestamps from block template

### PR DESCRIPTION
**Pull request by zeptin**
https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/4274

Basically, we were including transactions in the block template with timestamps greater than the coinstake transaction. Then this would happen:

1. Coinstake kernel gets found
2. Block gets stripped of all transactions with unusable timestamps
3. Fee/reward was not getting updated (and I'm not sure if it can be at this late stage, given that the coinstake is already signed? Maybe I'm wrong about this, I'll keep looking into it)

```                    // Make sure coinstake would meet timestamp protocol
                    // as it would be the same as the block timestamp.
                    block.Transactions[0].Time = block.Header.Time = coinstakeContext.CoinstakeTx.Time;

                    // We have to make sure that we have no future timestamps in
                    // our transactions set.
                    for (int i = block.Transactions.Count - 1; i >= 0; i--)
                    {
                        if (block.Transactions[i].Time > block.Header.Time)
                        {
                            this.logger.LogDebug("Removing transaction with timestamp {0} as it is greater than coinstake transaction timestamp {1}.", block.Transactions[i].Time, block.Header.Time);
                            block.Transactions.Remove(block.Transactions[i]);
                        }
                    }

                    block.Transactions.Insert(1, coinstakeContext.CoinstakeTx);
                    block.UpdateMerkleRoot();

                    // Append a signature to our block.
                    ECDSASignature signature = coinstakeContext.Key.Sign(block.GetHash());

                    block.BlockSignature = new BlockSignature { Signature = signature.ToDER() };
```

4. Block would fail consensus validation due to the fees being too high, and would be rejected.

This feels like an inelegant solution, so this PR is open to comment & better suggestions.

### CODE REVIEW

**Comment by fassadlr** [comment]
@zeptin please fix the solution build 👍

**Comment by zeptin** [comment]
I have investigated this further and found the following:

1. The validity of a coinstake kernel does not depend on the coinstake's outputs at all.
2. The block creation sequence is:
a. Valid kernel is found
b. Coinstake has outputs added & is signed (this is where the fees + reward get initially added)
c. Coinstake inserted into block
d. Block merkle root recomputed
e. Block is signed

So the approach in this PR is unnecessary. We should just recompute the fees after removing the unwanted transactions , then repeat steps (b) to (e) without invalidating the kernel. I will update the PR accordingly. It is somewhat non-trivial though, as we only have the transactions themselves available at this stage. We may need to propagate a txId <-> feeAmount mapping down through the PoS minter.

**Comment by dangershony** [comment]
What is the issue that is being fixed, I don't have access to the private link in the PR (why not make an GitHub issue of it, is it so secret?)

Any logic fixed is best to check what X is doing first.

Am I correct in my understanding that because coinstake walks back the time parameter to find a valid kernal we may include TRX that are now in the future in comparison to the kernal time?

**Comment by fassadlr** [comment]
@zeptin, @dangershony pointed out that we should look at perhaps removing the transactions with new timestamps *before* we sign the coinstake tx. So in the `CreateCoinstakeAsync` method, before we call `SignTransactionInput` we can do the check there? Thoughts?

File: src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs [hunk 1]
```diff
@@ -567,18 +568,53 @@ private async Task<bool> StakeAndSignBlockAsync(List<UtxoStakeDescription> utxoS
...
+                    // We have to make sure that we have no future timestamps in our transactions set.
+                    bool removed = false;
                     for (int i = block.Transactions.Count - 1; i >= 0; i--)
                     {
                         if (block.Transactions[i].Time > block.Header.Time)
                         {
+                            Money feeToRemove = blockTemplate.FeeDetails[block.Transactions[i].GetHash()];
+
+                            // Remove the fee from a suitable coinstake output.
+                            for (int j = 0; j < coinstakeContext.CoinstakeTx.Outputs.Count; j++)
```
[the remark this answers is not in the document]
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs]: With the removal logic shuffled around this is no longer necessary, the split will happen after fee correction

File: src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs [hunk 2]
```diff
@@ -567,18 +568,53 @@ private async Task<bool> StakeAndSignBlockAsync(List<UtxoStakeDescription> utxoS
...
+                            {
+                                if (coinstakeContext.CoinstakeTx.Outputs[j].Value >= feeToRemove)
```
[the remark this answers is not in the document]
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs]: No longer applicable

File: src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs [hunk 3]
```diff
@@ -567,18 +568,53 @@ private async Task<bool> StakeAndSignBlockAsync(List<UtxoStakeDescription> utxoS
...
                             this.logger.LogDebug("Removing transaction with timestamp {0} as it is greater than coinstake transaction timestamp {1}.", block.Transactions[i].Time, block.Header.Time);
                             block.Transactions.Remove(block.Transactions[i]);
+                            removed = true;
+                        }
+                    }
+
+                    // We removed at least one transaction, so now the coinstake output that contains the fees is most likely inaccurate. Sign the coinstake again.
+                    if (removed)
+                    {
+                        try
```
[the remark this answers is not in the document]
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs]: It had extremely awkward parameter requirements, but in any case, this code is no longer necessary

**Comment by zeptin** [comment]
> @zeptin, @dangershony pointed out that we should look at perhaps removing the transactions with new timestamps _before_ we sign the coinstake tx. So in the `CreateCoinstakeAsync` method, before we call `SignTransactionInput` we can do the check there? Thoughts?

Yeah, I've done the modification and it flows much better 👍

**Comment by dangershony** [comment]
Just to be sure, I'd suggest to find where in X this is done

**Pull request merged**